From d24c0af818a35410f182757305eda8164f3c90bd Mon Sep 17 00:00:00 2001
From: Sultanxda <sultanxda@gmail.com>
Date: Sat, 6 May 2017 00:58:11 -0700
Subject: [PATCH] Eleven: Fix saving/restoring user settings

*Fix saved repeat-mode and shuffle-mode settings not being restored
*Fix shake-to-play and show-album-artwork settings not being saved
*Fix saving and restoring music visualizer setting

Change-Id: I7ebf1d3e17c00236eb2dd6cf504a76f1dba4ee19
---
 src/com/cyanogenmod/eleven/IElevenService.aidl     |  1 +
 .../cyanogenmod/eleven/MusicPlaybackService.java   | 26 ++++++++++++++++++++++
 .../eleven/ui/activities/SettingsActivity.java     |  2 ++
 src/com/cyanogenmod/eleven/utils/MusicUtils.java   | 12 ++++++++++
 4 files changed, 41 insertions(+)

diff --git a/src/com/cyanogenmod/eleven/IElevenService.aidl b/src/com/cyanogenmod/eleven/IElevenService.aidl
index e3b0acf..06a3415 100644
--- a/src/com/cyanogenmod/eleven/IElevenService.aidl
+++ b/src/com/cyanogenmod/eleven/IElevenService.aidl
@@ -50,4 +50,5 @@ interface IElevenService
     int getAudioSessionId();
     void setShakeToPlayEnabled(boolean enabled);
     void setLockscreenAlbumArt(boolean enabled);
+    void setVisualizer(boolean enabled);
 }
diff --git a/src/com/cyanogenmod/eleven/MusicPlaybackService.java b/src/com/cyanogenmod/eleven/MusicPlaybackService.java
index 4fbbc63..d121064 100644
--- a/src/com/cyanogenmod/eleven/MusicPlaybackService.java
+++ b/src/com/cyanogenmod/eleven/MusicPlaybackService.java
@@ -665,6 +665,9 @@ public void onCreate() {
                 PreferenceUtils.SHOW_ALBUM_ART_ON_LOCKSCREEN, true);
         setShakeToPlayEnabled(mPreferences.getBoolean(PreferenceUtils.SHAKE_TO_PLAY, false));
 
+        mRepeatMode = mPreferences.getInt("repeatmode", REPEAT_NONE);
+        mShuffleMode = mPreferences.getInt("shufflemode", SHUFFLE_NONE);
+
         registerExternalStorageListener();
 
         // Initialize the media player
@@ -2866,6 +2869,9 @@ public void setShakeToPlayEnabled(boolean enabled) {
         if (D) {
             Log.d(TAG, "ShakeToPlay status: " + enabled);
         }
+        final SharedPreferences.Editor editor = mPreferences.edit();
+        editor.putBoolean(PreferenceUtils.SHAKE_TO_PLAY, enabled);
+        editor.apply();
         if (enabled) {
             if (mShakeDetector == null) {
                 mShakeDetector = new ShakeDetector(mShakeDetectorListener);
@@ -2885,10 +2891,22 @@ public void setShakeToPlayEnabled(boolean enabled) {
      */
     public void setLockscreenAlbumArt(boolean enabled) {
         mShowAlbumArtOnLockscreen = enabled;
+        final SharedPreferences.Editor editor = mPreferences.edit();
+        editor.putBoolean(PreferenceUtils.SHOW_ALBUM_ART_ON_LOCKSCREEN, enabled);
+        editor.apply();
         notifyChange(META_CHANGED);
     }
 
     /**
+     * Called to save visibility of visualizer in app
+     */
+    public void setVisualizer(boolean enabled) {
+        final SharedPreferences.Editor editor = mPreferences.edit();
+        editor.putBoolean(PreferenceUtils.SHOW_VISUALIZER, enabled);
+        editor.apply();
+    }
+
+    /**
      * Called to start listening to shakes
      */
     private void startShakeDetector() {
@@ -3851,6 +3869,14 @@ public void setLockscreenAlbumArt(boolean enabled) {
             mService.get().setLockscreenAlbumArt(enabled);
         }
 
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public void setVisualizer(boolean enabled) {
+            mService.get().setVisualizer(enabled);
+        }
+
     }
 
     private class QueueUpdateTask extends AsyncTask<Void, Void, List<MediaSession.QueueItem>> {
diff --git a/src/com/cyanogenmod/eleven/ui/activities/SettingsActivity.java b/src/com/cyanogenmod/eleven/ui/activities/SettingsActivity.java
index 856e185..bd14df8 100644
--- a/src/com/cyanogenmod/eleven/ui/activities/SettingsActivity.java
+++ b/src/com/cyanogenmod/eleven/ui/activities/SettingsActivity.java
@@ -109,6 +109,8 @@ public void onSharedPreferenceChanged(SharedPreferences sharedPreferences,
             MusicUtils.setShakeToPlayEnabled(sharedPreferences.getBoolean(key, false));
         } else if (key.equals(PreferenceUtils.SHOW_ALBUM_ART_ON_LOCKSCREEN)) {
             MusicUtils.setShowAlbumArtOnLockscreen(sharedPreferences.getBoolean(key, true));
+        } else if (key.equals(PreferenceUtils.SHOW_VISUALIZER)) {
+            MusicUtils.setShowVisualizer(sharedPreferences.getBoolean(key, true));
         }
     }
 }
diff --git a/src/com/cyanogenmod/eleven/utils/MusicUtils.java b/src/com/cyanogenmod/eleven/utils/MusicUtils.java
index 8ab4c73..1554e5b 100644
--- a/src/com/cyanogenmod/eleven/utils/MusicUtils.java
+++ b/src/com/cyanogenmod/eleven/utils/MusicUtils.java
@@ -299,6 +299,18 @@ public static void setShowAlbumArtOnLockscreen(final boolean enabled) {
     }
 
     /**
+     * Set show visualizer in app
+     */
+    public static void setShowVisualizer(final boolean enabled) {
+        try {
+            if (mService != null) {
+                mService.setVisualizer(enabled);
+            }
+        } catch (final RemoteException ignored) {
+        }
+    }
+
+    /**
      * Changes to the next track asynchronously
      */
     public static void asyncNext(final Context context) {
